includes/class-sensei-course-structure.php

```diff
@@ -151,17 +151,331 @@ private function get_modules() : array {
+					return false;
+				}
+
+				list( $module_id, $module_lesson_ids ) = $save_module_result;
+
+				$lesson_ids     = array_merge( $lesson_ids, $module_lesson_ids );
+				$module_order[] = $module_id;
+
+			} elseif ( 'lesson' === $item['type'] ) {
+				$lesson_id = $this->save_lesson( $item );
+				if ( ! $lesson_id ) {
+					return false;
+				}
+
+				$lesson_ids[]   = $lesson_id;
+				$lesson_order[] = $lesson_id;
+
+				update_post_meta( $item['id'], '_order_' . $this->course_id, count( $lesson_ids ) );
+			}
+		}
+
+		// Save the module association.
+		$module_diff = array_diff( $current_module_ids, $module_order );
+		if ( ! empty( $module_diff ) || count( $current_module_ids ) !== count( $module_order ) ) {
+			wp_set_object_terms( $this->course_id, $module_order, 'module' );
+		}
+
+		// Save the module order.
+		$this->save_module_order( $module_order );
+
+		// Save the lesson order.
+		update_post_meta( $this->course_id, '_lesson_order', implode( ',', $lesson_order ) );
+
+		// Delete removed modules and lessons.
+		$delete_lesson_ids = array_diff( $current_lesson_ids, $lesson_ids );
+		foreach ( $delete_lesson_ids as $lesson_id ) {
+			$this->clear_lesson_associations( $lesson_id );
+		}
+
+		delete_transient( 'sensei_' . $this->course_id . '_none_module_lessons' );
+
+		return true;
+	}
+
+	/**
+	 * Save a module item.
+	 *
+	 * @param array $item Item to save.
+	 *
+	 * @return false|array[] {
+	 *     If successful, we return this:
+	 *
+	 *     @type int   $0 $module_id  Saved module ID.
+	 *     @type array $1 $lesson_ids All the lesson IDs from this module.
+	 * }
+	 */
+	private function save_module( array $item ) {
+		if ( $item['id'] ) {
+			$module_id = $this->update_module( $item );
+		} else {
+			$module_id = $this->create_module( $item );
+		}
+
+		if ( ! $module_id ) {
+			return false;
+		}
+
+		$lesson_ids       = [];
+		$lesson_order_key = '_order_module_' . $module_id;
+		foreach ( $item['lessons'] as $lesson_item ) {
+			$lesson_id = $this->save_lesson( $lesson_item, $module_id );
+			if ( ! $lesson_id ) {
+				return false;
+			}
+
+			wp_set_object_terms( $lesson_id, [ $module_id ], 'module' );
+			update_post_meta( $lesson_id, $lesson_order_key, count( $lesson_ids ) );
+			delete_post_meta( $lesson_id, '_order_' . $this->course_id );
+
+			$lesson_ids[] = $lesson_id;
+		}
+
+		return [
+			$module_id,
+			$lesson_ids,
+		];
+	}
+
+	/**
+	 * Create a module.
+	 *
+	 * @param array $item Item to create.
+	 *
+	 * @return false|int
+	 */
+	private function create_module( array $item ) {
+		$args = [
+			'description' => $item['description'],
+		];
+
+		$teacher_user_id = get_post( $this->course_id )->post_author;
+		if ( ! user_can( $teacher_user_id, 'manage_options' ) ) {
+			$args['slug'] = intval( $teacher_user_id ) . '-' . sanitize_title( $item['title'] );
+		}
+
+		$create_result = wp_insert_term( $item['title'], 'module', $args );
+		if ( is_wp_error( $create_result ) ) {
+			return false;
+		}
```

Comment:
This throws a `term_exists` error when there is already a module with that name. Not sure if we can get around that, but if not, we should get the error message to the user somehow. 

Best bet is returning the `WP_Error` everywhere, but for clarity, we could prefix the error message to something like `Module "Getting Started": [originalMessage]` (or add that as an extra field and do it on the client)

Reply:
After chatting I set it up to use existing modules available to the user dd8a1f798322fa6a06ce16feeb4464486416d4b7

I also added protection against the user attempting to add two modules with the same name (that will pass back the `sensei_course_structure_duplicate_items` error)